File: 1-two-sum/two-sum.scala
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        nums.zipWithIndex
            .foldLeft((Map.empty[Int, Int], Option.empty[Array[Int]])) {
                case ((seen, result @ Some(_)), _) => (seen, result) // result already found
                case ((seen, None), (num, idx)) =>
                    val complement = target - num
                    seen.get(complement) match {
                        case Some(prevIdx) => (seen, Some(Array(prevIdx, idx)))
                        case None => (seen + (num -> idx), None)
                    }
            }._2.getOrElse(throw new IllegalArgumentException("No two sum solution"))
    }
}

File: 1-two-sum/two-sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    Enum.reduce_while(Enum.with_index(nums), %{}, fn {num, idx}, seen ->
      complement = target - num
      
      case Map.get(seen, complement) do
        nil ->
          {:cont, Map.put(seen, num, idx)}
        prev_idx ->
          {:halt, [prev_idx, idx]}
      end
    end) || raise "No two sum solution"
  end
end